test/core/client_channel/resolvers/custom_resolver_test.cc

```diff
@@ -0,0 +1,359 @@
+  }
+
+ private:
+  grpc_resolver_observer* observer_;
+  grpc_addresses* reresolution_response_ = nullptr;
+};
+
+class ResolverFactory {
+ public:
+  ResolverFactory() {}
+  virtual ~ResolverFactory() {}
+
+  virtual Resolver* Resolve(grpc_resolver_args* args,
+                            grpc_resolver_observer* observer) = 0;
+
+  static void* ResolveWrapper(void* user_data, grpc_resolver_args* args,
+                              grpc_resolver_observer* observer) {
+    return static_cast<ResolverFactory*>(user_data)->Resolve(args, observer);
+  }
+  static void DestroyWrapper(void* user_data) {
+    delete static_cast<ResolverFactory*>(user_data);
+  }
+};
+
+class FakeResolver : public Resolver {
+ public:
+  FakeResolver(grpc_resolver_args* args, grpc_resolver_observer* observer)
+      : Resolver(observer) {}
+
+  void RequestReresolution() {
+    if (reresolution_response_ != nullptr) {
+      SetAddresses(reresolution_response_);
+    }
+  }
+
+  void SetResponse(const Addresses& addresses) { SetAddresses(addresses); }
+
+  void SetReresolutionResponse(Addresses addresses) {
+    reresolution_response_ = std::move(addresses);
+  }
+
+ private:
+  Addresses reresolution_response_ = nullptr;
+};
+
+class FakeResolverFactory : public ResolverFactory {
+ public:
+  FakeResolverFactory(FakeResolver** resolver) : resolver_(resolver) {}
+
+  Resolver* Resolve(grpc_resolver_args* args,
+                    grpc_resolver_observer* observer) {
+    GPR_ASSERT(resolver_ != nullptr);
+    auto resolver = new FakeResolver(args, observer);
+    *resolver_ = resolver;
+    resolver_ = nullptr;
+    return resolver;
+  }
+
+ private:
+  FakeResolver** resolver_;
+};
+
+static grpc_core::OrphanablePtr<grpc_core::Resolver> build_custom_resolver(
+    const char* scheme, grpc_combiner* combiner) {
+  grpc_core::ResolverFactory* factory =
+      grpc_core::ResolverRegistry::LookupResolverFactory(scheme);
+  grpc_channel_args channel_args = {0, nullptr};
+  grpc_core::ResolverArgs args;
+  args.args = &channel_args;
+  args.combiner = combiner;
+  return factory->CreateResolver(args);
+}
+
+static void register_resolver_factory(const char* scheme,
+                                      ResolverFactory* factory) {
+  grpc_resolver_factory_register(
+      scheme, factory, ResolverFactory::ResolveWrapper,
+      ResolverFactory::DestroyWrapper, Resolver::RequestReresolutionWrapper,
+      Resolver::DestroyWrapper, nullptr);
+}
+
+TEST(CustomResolverTest, End2End) {
+  FakeResolver* fake_resolver = nullptr;
+  register_resolver_factory("custom", new FakeResolverFactory(&fake_resolver));
+
+  grpc_core::ExecCtx exec_ctx;
+  grpc_combiner* combiner = grpc_combiner_create();
+  // Create resolver.
+  grpc_core::OrphanablePtr<grpc_core::Resolver> resolver =
+      build_custom_resolver("custom", combiner);
+  GPR_ASSERT(resolver.get() != nullptr);
+  GPR_ASSERT(fake_resolver != nullptr);
+  // Test 1: normal resolution.
```

Comment:
It looks like most of this was copied from fake_resolver_test.cc.  However, it's not clear to me that these tests make sense for the custom resolver -- we can probably do something much simpler here, because there's less state than there is in the fake resolver.

I think all we really need here is a couple of basic checks:

1. When the resolver plugin returns results to the observer, those results are returned to the client channel via `CustomResolver::NextLocked()`.

2. When `CustomResolver::RequestReresolutionLocked()` is called, the plugin's `request_reresolution()` method is called.